src/emitEventsInBatches.js

```diff
@@ -0,0 +1,55 @@
+const chunk = require('lodash.chunk');
+const omit = require('lodash.omit');
+const { enrichMeta, partitionKey } = require('./enrich')
+
+const MAX_RECORDS = 500; 
+
+const enrichRecords = (events, config) => {
+  return events.map(event => {
+    const enrichedEvent = enrichMeta(event, config.appName, config.ipv4);
+    return {
+      Data: JSON.stringify(enrichedEvent),
+      PartitionKey: config.partitionKey || partitionKey(enrichedEvent),
+    }
+  })
+};
+
+const emitEvents = async (kinesis, records, config, retries) => {
+  const params = {
+    Records: records,
+    StreamName: config.kinesisStream.resource
+  };
+
+  try {
+    const { FailedRecordCount, Records } = await kinesis.putRecords(params).promise();
+
+    if(FailedRecordCount !== 0) {
+      const failedEvents = [];
+      Records.forEach((failedRecord, index) => {
+        if (failedRecord.ErrorCode) {
+          failedEvents.push({
+            failedRecord: records[index],
+            failedEvent: omit(JSON.parse(records[index].Data), ['meta']),
```

Comment:
Not sure what are the use cases for this `failedEvent` so hard to tell, but I think this operation might not necessarily be safe, as it seems like it could be valid for consumers to put things in `meta`.
I think if the goal is really to have as `failedEvent` the exact event that was originally sent, this function should take both `records` and `events` in, and find back the event here with `events[index]`.
But again, maybe it's unimportant that the object changed in the process, not sure.

Reply:
we talked offline and resolved this!